certmanager/generate_cert_test.go

```diff
@@ -0,0 +1,234 @@
+				Org: "",
+				IsCA: false,
+				IsSelfSigned: false,
+				IsClient: true,
+			},
+		},
+		{
+			// server cert with IP as SAN
+			certOptions: CertOptions{
+				Host: "1.2.3.4",
+				ValidFrom: nowInString,
+				ValidFor: time.Hour * 24,
+				SignerCert: caCert,
+				SignerPriv: caPriv,
+				Org: "",
+				IsCA: false,
+				IsSelfSigned: false,
+				IsClient: false,
+			},
+		},
+		{
+			// client cert with service account as SAN
+			certOptions: CertOptions{
+				Host: "istio:foo.serviceaccount.com",
+				ValidFrom: nowInString,
+				ValidFor: time.Hour * 100,
+				SignerCert: caCert,
+				SignerPriv: caPriv,
+				Org: "",
+				IsCA: false,
+				IsSelfSigned: false,
+				IsClient: true,
+			},
+		},
+		{
+			// server cert with service account as SAN
+			certOptions: CertOptions{
+				Host: "istio:bar.serviceaccount.com",
+				ValidFrom: nowInString,
+				ValidFor: time.Hour * 50,
+				SignerCert: caCert,
+				SignerPriv: caPriv,
+				Org: "",
+				IsCA: false,
+				IsSelfSigned: false,
+				IsClient: false,
+			},
+		},
+	}
+	for _, c := range cases {
+		certOptions := c.certOptions
+		certPem, privPem := GenCert(certOptions)
+		notAfter = notBefore.Add(certOptions.ValidFor)
+
+		var extKeyUsage x509.ExtKeyUsage
+		if certOptions.IsClient {
+			extKeyUsage = x509.ExtKeyUsageClientAuth
+		} else {
+			extKeyUsage = x509.ExtKeyUsageServerAuth
+		}
+		verifyCert(t, privPem, certPem, caCertPem, certOptions.Host, VerifyFields{
+			notBefore: notBefore,
+			notAfter: notAfter,
+			extKeyUsage: extKeyUsage,
+			isCA: false,
+			org: "MyOrg",
+		})
+	}
+}
+
+func verifyCert(t *testing.T, privPem []byte, certPem []byte, rootCertPem []byte, host string, expectedFields VerifyFields) {
+	roots := x509.NewCertPool()
+	var ok bool
+	if rootCertPem == nil {
+		ok = roots.AppendCertsFromPEM(certPem)
+	} else {
+		ok = roots.AppendCertsFromPEM(rootCertPem)
+	}
+
+	if !ok {
+		t.Errorf("failed to parse root certificate")
+	}
+
+	block, _ := pem.Decode(certPem)
+	if block == nil {
+		t.Errorf("failed to parse certificate PEM")
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		t.Errorf("failed to parse certificate: " + err.Error())
+	}
+
+	san := host
+	// uri scheme is currently not supported in go VerifyOptions. We verify
+	// this uri at the end as a special case.
+	if strings.HasPrefix(host, uriScheme) {
+		san = ""
+	}
+	opts := x509.VerifyOptions{
+		DNSName:	san,
+		Roots:		roots,
+	}
+	opts.KeyUsages = append(opts.KeyUsages, x509.ExtKeyUsageAny)
+
+	if _, err := cert.Verify(opts); err != nil {
+		t.Errorf("failed to verify certificate: " + err.Error())
+	}
+
+	block, _ = pem.Decode(privPem)
+	if block == nil || block.Type != "RSA PRIVATE KEY" {
+		t.Errorf("failed to decode PEM block containing private key")
+	}
+	priv, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		t.Errorf("failed to parse private key: " + err.Error())
+	}
+	if !reflect.DeepEqual(priv.PublicKey, *cert.PublicKey.(*rsa.PublicKey)) {
```

Comment:
You probably don't need to type conversion for `cert.PublicKey`. I think `reflect.DeepEqual()` is smart enough to detect that.